### PR TITLE
Add Retro Shift (Auto Shift for Tap Hold via Retro Tapping) and Custom Auto Shifts

### PR DESCRIPTION
**Pull request by IsaacElenbaas**
This pull request will allow one to use Auto Shift (AS) on a key but also use it as a modifier - three keys in one, and it feels like they're completely separate! I've been using a very janky hardcoded version of this for months, it's great. It will also add support for custom AS keys, as things like `.` and `!` are clearly a better match than those they currently have.

## Description

This is a recode of https://github.com/qmk/qmk_firmware/pull/9889, as upon further testing it didn't work when `TAPPING_TERM` was not exceeded (and Auto Shift now supports keyrepeat).

https://github.com/p00ya/qmk_firmware/pull/2 has also been merged with this pull as there are quite a few conflicts between the two features. I also don't want to have to manage those conflicts once I've forgotten how AS works (I had a custom shifts pull open for months and then helped completely rewrite Auto Shift and got that merged first).

`get_custom_auto_shifted_key` was added, as `get_auto_shifted_key` would result in a lot of duplicate and possibly outdated code (e.g. if a `#define` is added to Retro Shift everything).

`AUTO_SHIFT_TIMEOUT_PER_KEY`, `AUTO_SHIFT_REPEAT_PER_KEY`, and `AUTO_SHIFT_NO_AUTO_REPEAT_PER_KEY` were also added, as they are nearly a necessity with the many different purposes custom shifted keys can have.

`HOLD_ON_OTHER_KEY` is already handled, so if that is merged it's not an issue.

## Known Bugs/Inconsistencies/Todo

Retro Shift:
* None

Custom Shifts:
* None

## Bugfixes
Fixed rolling from one AS key to another and holding the second sending both unshifted
Stopped tapping an AS key, tapping a non-AS key, and holding the same AS key from keyrepeating
Alternate #12083, as it won't work for Custom Shifts

## Types of Changes

- [x] Core
- [X] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by IsaacElenbaas** [comment]
@manna-harbour It would be great if you could look over what I have so far before it gets polluted with custom shifts. I have to assume you know more about `action.c` and `action_tapping.c` than me, seeing as you made the original pull and my total experience is from today haha

**Comment by IsaacElenbaas** [comment]
Oh, and the docs say "Auto Shift will be enabled automatically if it is not already" - I don't know how to do it, but I'm sure it can be done. If it can't that will need to be removed.

**Comment by IsaacElenbaas** [comment]
[`is_tap_action`](https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/action.c#L1046) may be what I should be using instead of those long ifs. Not sure if they're equivalent, and if not, which should be used, but Retro Tapping uses `is_tap_action`.

**Comment by IsaacElenbaas** [comment]
If anyone is testing `TAPPING_TERM`, `PERMISSIVE_HOLD`, and `IGNORE_MOD_TAP_INTERRUPT`, [this](https://cdn.discordapp.com/attachments/451392568345755648/747388373781708930/tap-holds-qmk.pdf) may be helpful. `IGNORE_MOD_TAP_INTERRUPT` may appear to be broken, as it always sends both keys lowercased regardless of the second's release time, but this is an AS bug (rolling and holding the second of two AS keys does the same). It should be a one-line fix, I'll test later and add to this.

**Comment by IsaacElenbaas** [comment]
I have fixes for the "Keyrepeating a Tap Hold key, then tapping any other key any number of times" bug, possibly the mixing shifts weirdness, two AS-core bugs, and will fix `IGNORE_MOD_TAP_INTERRUPT` not releasing the modifier sometimes. Should be up tonight or tomorrow.

**Comment by IsaacElenbaas** [comment]
Alright, so I'm changing up how `IGNORE_MOD_TAP_INTERRUPT` and `PERMISSIVE_HOLD` work in the next commit. You can check the docs once it's up for more info, but IMTI only applies to rollover, PH has no effect, and all nested presses trigger hold action. It is mostly all because holding for longer to exceed `TAPPING_TERM` makes no sense as your release times should be muscle memory based on whether or not you want that *shifted,* and lots of Tap Hold config is to change behavior based on TT. Making a Tap Hold a non-AS key once customs are added makes it use normal config behavior (I've already tested that). I would put this behind a `#define`, but I can't find how to tell (if there is a way) whether a Tap Hold record is going to be the hold action, as they're sent in the physical order regardless of whether that's 'corrected' by your settings. Unless I have that, it's a whole program flow recode to switch it.

**Comment by IsaacElenbaas** [comment]
There are definitely going to be quirks to fix when mixing Retro Shifts with non-AS keys, but otherwise this is back to complete in regards to Retro Shift.

**Comment by IsaacElenbaas** [comment]
We're closing in on Retro Shift being stable, and then hopefully I can make the code less horrible.
* Holding a Layer Tap and an Auto Shift key past `TAPPING_TERM`, then releasing the Layer Tap first leaves the layer on and sends the key on the new layer in the place of the Auto Shift key shifted. Neither should happen. `IMTI` set but no `PERMISSIVE` or `HOLD_ON_OTHER`, I haven't tested with other configs.
* Currently it is impossible to send a shifted Auto Shift key using a Layer Tap until the layer action was decided. This should be a relatively easy fix, I just need to store something like `retroshift_time` for the first tap after a Retro Shift down and only reference it on a release that triggers a Retro evaluation.
* Non-AS key down events will *always* evaluate a Tap Hold key. This means that one cannot roll over with a Tap Hold and non-AS key with `IMTI` to get both their tap actions. Instead of "fixing" this, as it would cause Auto Shift to delay down events of non-AS keys, I need to add a docs section saying to just return true for that (or every) key in `is_custom` so that they are handled by AS and check physical shift state when pressing them and undo weak shift.

While working on those bugfixes `action_tapping`'s effects became quite noticeable. As I am reworking Tap Hold configurations to not be timing-based, it might be best to call some AS function to possibly handle Tap Holds before `action_tapping` to prevent, say, keycodes being delayed until `TAPPING_TERM` unnecessarily. I'm not sure if keycodes are available early enough, I'll think on it.

**Comment by IsaacElenbaas** [comment]
I plan to rework this to play more nicely with the rest of QMK and hopefully finalize everything + merge custom shifts over Spring Break (by the 13th).

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

File: quantum/process_keycode/process_auto_shift.c
Comment (IsaacElenbaas) [inline quantum/process_keycode/process_auto_shift.c:224]: This is horrible but for some reason I can't think of a better way to do it. Send help.

**Comment by IsaacElenbaas** [comment]
`NO_ACTION_TAPPING` guards should probably be added to all the `RETRO_SHIFT` checks.

**Comment by IsaacElenbaas** [comment]
I'm not sure how we want to merge this and #11536, or whether `autoshift_is_custom` is convenient enough to just not merge that.

File: quantum/process_keycode/process_auto_shift.c
Comment (IsaacElenbaas) [reply quantum/process_keycode/process_auto_shift.c]: It's gotten worse but due to the `_per_key` stuff it definitely can't get better unfortunately.

**Comment by IsaacElenbaas** [comment]
I've just realized I'm still using an array of bits. Not the best solution memory-wise, but C doesn't have bitmaps, and storing in an integer array would get messy. . .

File: tmk_core/common/action.c
Comment (IsaacElenbaas) [inline tmk_core/common/action.c:46]: I don't believe the `NO_ACTION_TAPPING` checks here or just below are necessary.

**Comment by mrlinuxfish** [comment]
Retro Tapping appears to break with `#define IGNORE_MOD_TAP_INTERRUPT` in `config.h`. If I use retro tapping with this setting it outputs `enter d` instead of `#` on my symbol layer. This is the tap key on the layer tap followed by the key on the base layer. `LT(_SYM, KC_ENT)`

**Comment by IsaacElenbaas** [comment]
Fixed. Not related to `IGNORE_MOD_TAP_INTERRUPT`, I had seen that Mod Taps by default had nested presses under `TAPPING_TERM` act like holds and didn't realize Layer Taps did not. Last commit also fixed `PERMISSIVE_HOLD` and `TAPPING_TERM` `_PER_KEY` possibly breaking Retro Shift behavior.

**Comment by IsaacElenbaas** [comment]
There's waaaaay too much `#define` hell in this PR for my liking.

**Comment by IsaacElenbaas** [comment]
I've just realized that I'm using `register_code` and not `register_code16` everywhere. For custom shifts this is a problem. At the very least, the documentation example and `autoshift_press_user` default need to be changed, but they will still need testing. Not sure when I'll get time to look into that.

**Comment by IsaacElenbaas** [comment]
> LGUI mod tap + LCTL mod tap + autoshift key causes the autoshift key to always (?) be shifted.

Holding the `LGUI` MT past `TAPPING_TERM` before pressing the autoshift key or releasing only the `LCTL` quickly makes it work properly. When the bug occurs, nothing is sent for some time. The culprit could be [this line,](https://github.com/IsaacElenbaas/qmk_firmware/blob/as_custom_shifts_retro/tmk_core/common/action_tapping.c#L118) which is there so that the hold action is not selected until after `RETRO_SHIFT` is exceeded even if it is greater than `TAPPING_TERM`.

Also, in glancing at this problem, I noticed that a few places checking `IS_RETRO` don't check `get_retro_tapping`. Those need to be added.

**Comment by IsaacElenbaas** [comment]
The `get_retro_tapping` checks were there, just in an odd spot that made the code hard to read. Moved. Last commit also made keys send as soon as the action was determined, which A: is great and B: should prevent any other weird cases like the chained MTs one it fixed.

File: quantum/process_keycode/process_auto_shift.c
Comment (IsaacElenbaas) [reply quantum/process_keycode/process_auto_shift.c]: That one is nothing compared to `action_tapping.c`. . . the main if now takes up my entire screen.

File: quantum/process_keycode/process_auto_shift.h
Comment (IsaacElenbaas) [inline quantum/process_keycode/process_auto_shift.h:34]: Maintainers, thoughts? Ugly, and not very C-like, but this pull isn't a breaking change at the moment and I would like to keep it that way. I also don't want Auto Shift to have the single `PER_KEY` option that isn't called `get_[value]`. This makes `get_autoshift_timeout` either call the user function, or `get_generic_autoshift_timeout` if not passed the keycode and record pointer. The only impact on the user is that their `get_autoshift_timeout` definition must have parentheses around it (noted in the docs).

File: quantum/process_keycode/process_auto_shift.h
Comment (IsaacElenbaas) [inline quantum/process_keycode/process_auto_shift.h:40]: This allows `case AUTO_SHIFT_ALPHA:` and the like in the user's `get_autoshift_timeout`. Janky, but I think it should be like this.

**Comment by IsaacElenbaas** [comment]
Please don't, stale-bot.
 * The large `if` in `action_tapping.c` may be a bit *too* large, though I don't see how I would change that
 * There are the two wack `#define` things I did in `process_auto_shift.h`
 * The docs say "Auto Shift will be enabled automatically if it is not already" and I do not know how to do that (checking `define` in makefile), though I assume it is possible - otherwise that should be removed
 * `action.c` needs a rebase, but it's very obvious what is not mine - I will try to remember to do that later
 
Other than the above, I feel this is ready to be merged. ~5 other people have used / are still using this, and I have been using it on all of my boards for ~4 months. I am using everything in this pull and everything that should affect it except one shots.

File: docs/tap_hold.md
Comment (drashna) [inline docs/tap_hold.md:205]: If this is using functions from `process_auto_shift` then that's a problem

File: docs/tap_hold.md
Comment (drashna) [inline docs/tap_hold.md:212]: If this does use functions from `process_auto_shift`, then this should be a make/rules thing, not a define.

File: quantum/process_keycode/process_auto_shift.h
Comment (drashna) [reply quantum/process_keycode/process_auto_shift.h:34]: Definitely ugly.

This should be a c function, rather than a define.

File: quantum/process_keycode/process_auto_shift.h
Comment (drashna) [reply quantum/process_keycode/process_auto_shift.h:40]: I don't agree.  This removes a lot of code and obfuscates it, making it less readable/understandable.

**Review by drashna — COMMENTED** [review]
To be honest, it feels like some of this code should be pulled out and moved into a separate handler, rather than the additional code added on. 

There is definite overlap with autoshift, but including the files from it doesn't feel right, in this context.

File: docs/tap_hold.md
Comment (IsaacElenbaas) [reply docs/tap_hold.md:212]: Why? Permissive hold, ignore mod tap interrupt, and even the 'stock' retro tapping are all `#defines`. This is just retro tapping for Auto Shift, and only in its own `#define` because it makes the hold have *two* possible actions. This is not an alternative to Auto Shift's normal functionality.

Comment (drashna) [reply docs/tap_hold.md:212]: > If this does use functions from `process_auto_shift`

If it's actually using functions from the C file, then it needs to be sourced, too.  And if that's the case, the way to do that is via the make system.  

If it's just using defines from it, then ... I'm not sure they belong in the auto shift header file, then, and should probably go into the action_tapping.h file.

File: docs/tap_hold.md
Comment (IsaacElenbaas) [reply docs/tap_hold.md:205]: I'm not sure I follow. Do you mean that it should be a separate feature that can work without Auto Shift enabled?

File: docs/tap_hold.md
Comment (IsaacElenbaas) [reply docs/tap_hold.md:212]: It is *inside* `process_auto_shift`. As above, like retro tapping, this is a modifier for the (Auto Shift's) behavior with Tap Holds that are released without using their hold action on another key. Any references to the `#define` outside `process_auto_shift` (such as in `action.c` and `action_tapping.c`, as the hold action needs to be prevented sometimes because it cannot be captured and interpreted easily) also check if Auto Shift is enabled.

File: docs/tap_hold.md
Comment (drashna) [reply docs/tap_hold.md:205]: Wait, so this does require autoshift? 

If so, this documentation should be in that documentation, and not here.   Because otherwise, it sounds like this is separate functionality.  Hence my comments.

File: quantum/process_keycode/process_auto_shift.h
Comment (IsaacElenbaas) [reply quantum/process_keycode/process_auto_shift.h:34]: Alright, I will change it. I suppose something changing this much would probably go through the breaking change process, anyway (I believe the keyrepeat one did), and it's not like it's hard to quick change a single function name when it breaks people's keymaps.

File: docs/tap_hold.md
Comment (drashna) [reply docs/tap_hold.md:212]: Ah, okay. In that case, I would prefer that the documentation stay in the auto-shift doc, personally.   That way, it's less prone to confusion.   Maybe with a subnote about this functionality in autoshift.

File: docs/tap_hold.md
Comment (drashna) [inline docs/tap_hold.md:203]: 👍🏻

**Review by drashna — APPROVED** [review]

**Comment by IsaacElenbaas** [comment]
Well, that rebase was horrible - took me an hour and a half. Should be pushed in a second, I have one more thing to try. I kept trying to clean up the commit history, but editing anything before the merge commit makes it fail to apply (even after successful rebases). Its conflicts are because changes from commits prior to it weren't applied. No clue what's going on there, sorry to anyone actually looking through them. You can ignore everything before the one that says "rewrite." If I screwed something up, `a5b7ad881629c4c736dbe0049a46e00c7484097f` is pre-rebase.

**Comment by IsaacElenbaas** [comment]
heck

**Comment by IsaacElenbaas** [comment]
double heck

**Comment by IsaacElenbaas** [comment]
Somehow all commits got deleted, they're back up now on my branch but not in this - hopefully this will reopen and grab them properly?

**Comment by IsaacElenbaas** [comment]
The PR linter seems to disagree with my `qmk cformat`, though I have updated it.

**Comment by tzarc** [comment]
Just a heads-up, I intend to push this PR through this breaking changes cycle.

Unfortunately this PR changes areas of the code that have also had merges this breaking changes cycle, and I’m unsure of how the behaviour interacts… otherwise I’d resolve the conflicts myself. At this point it'd be great if you could get these conflicts sorted -- apologies if we create more conflicts during merges of other PRs — at this point the quicker they're sorted the more likely this makes it in first.

**Comment by IsaacElenbaas** [comment]
Those should be handled. No major conflicts, and nothing I wasn't expecting. Glancing over the overall diff everything looks fine. `HOLD_ON_OTHER_KEYPRESS` *should* already be handled here, but if it is broken it will only affect people trying to use both that and this feature and I would want to put it in a future pull anyway.

I tested this feature and with Auto Shift disabled again, but not with `HOLD_ON_OTHER_KEYPRESS` for the same reasons. The left half of my corne wouldn't send any keys, but I haven't touched anything near the split code (and haven't pulled in forever, I assume the split procedure has changed). Everything else seemed fine.

**Comment by IsaacElenbaas** [comment]
Doesn't have merge conflicts with #11036 but either will need to be updated when the other is merged.

**Comment by IsaacElenbaas** [comment]
@tzarc Conflicts fixed and I re-tested, working with early combos as well. Squashed this time as I hate every rebase taking forever to do. Notes above for #11036 still apply as that was delayed as well.

File: quantum/action_tapping.c
Comment (drashna) [inline quantum/action_tapping.c:121]: if both autoshift and retro shift are enabled, compilation errors occur

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: quantum/action_tapping.c
Comment (precondition) [inline quantum/action_tapping.c:127]: Is there a good reason why 0 is added to `RETRO_SHIFT` here and on line 363?

Comment (IsaacElenbaas) [reply quantum/action_tapping.c:127]: One can `#define RETRO_SHIFT` without a value, which just makes holds without interrupting keypresses send the shifted key. Defining it to a value makes mod taps' hold actions trigger after that many ms, like the [default behavior](https://cdn.discordapp.com/attachments/451392568345755648/747388373781708930/tap-holds-qmk.pdf) with `TAPPING_TERM` to allow using mod taps with the mouse.

Adding zero makes the former compile. I suppose those conditionals could not have the `+0` and be in another `#if`, but there are quite enough preprocessor macros in this haha

Comment (IsaacElenbaas) [reply quantum/action_tapping.c:127]: I assume you saw that when working on #15741 - I'll try to test that with Retro Shift (base Auto Shift won't have any issues) too soon, but if/when you have a go at it note that [you have to add mod/layer taps to Auto Shift.](https://github.com/qmk/qmk_firmware/issues/15458#issuecomment-1010448452)

**Pull request merged**
